test/jit/test_dtype_analysis.py

```diff
@@ -164,10 +177,59 @@ def test_binary_scalar(self):
                 scalar_type = in_dtypes[1]
 
                 if scalar_type == float32:
+
                     def add(x, y: float):
                         return x + y
+
                 else:
+
                     def add(x, y: int):
                         return x + y
 
                 self.assert_dtype_equal(add, in_shapes, in_dtypes)
+
+    def test_custom_rules(self):
+        # Test some of the ops that are not covered by Metatensors
+
+        # Note that unlike the Conv2d module, the function conv2d
+        # does not take dtype/device arguments.
+
+        def conv2d_fn(input, weight, bias):
+            return torch.nn.functional.conv2d(input, weight, bias)
+
+        def adaptive_avg_pool2d_fn(input, output_size: Tuple[int]):
+            return torch._C._nn.adaptive_avg_pool2d(input, output_size)
+
+        for fn, inputs_fn in (
+            (conv2d_fn, sample_inputs_conv2d),
+            (adaptive_avg_pool2d_fn, sample_inputs_adaptive_avg_pool2d),
+        ):
+            for dtype in (torch.int8, torch.float64):
+                # Gets default version for conv2d
+                sample_input: SampleInput = inputs_fn(None, "cpu", dtype, False)[-1]
+                input_args = [sample_input.input, *sample_input.args]
+                self.assert_dtype_equal_custom_args(fn, input_args)
+
+        # Now make sure that conv2d doesn't support mixed args
+        conv_ins = sample_inputs_conv2d(None, "cpu", torch.float, False)
+        conv_in = conv_ins[-1]
+        weight, bias = conv_in.args
+        weight = weight.type(torch.long)
+
+        with self.assertRaises(RuntimeError):
```

Comment:
can have a test to assert that dtype_propagation doesn't dtype the output tensor in this case?

Reply:
Makes sense. I'll have to add a bit more logic to handle that.